pkg/tbtc/maintainer/bitcoin_difficulty.go

```diff
@@ -2,54 +2,271 @@ package maintainer
+	if err != nil {
+		return fmt.Errorf(
+			"cannot check whether Bitcoin difficulty maintainer is authorized "+
+				"to submit block headers [%v]",
+			err,
+		)
+	}
+
+	if !isAuthorized {
+		return fmt.Errorf(
+			"bitcoin difficulty maintainer has not been authorized to submit " +
+				"block headers",
+		)
+	}
+
+	return nil
+}
+
+// proveNextEpoch proves a single Bitcoin blockchain epoch in the Bitcoin
+// difficulty chain if there is a Bitcoin blockchain epoch to be proven.
+func (bdm *BitcoinDifficultyMaintainer) proveNextEpoch() error {
+	// The height of the Bitcoin blockchain.
+	currentBlockNumber, err := bdm.btcChain.GetCurrentBlockNumber()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get current block number [%v]",
+			err,
+		)
+	}
+
+	// The current epoch proven in the Bitcoin difficulty chain.
+	currentEpoch, err := bdm.chain.CurrentEpoch()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get current epoch [%v]",
+			err,
+		)
+	}
+
+	// The number of blocks required for each side of a retarget proof.
+	proofLength, err := bdm.chain.ProofLength()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get proof length [%v]",
+			err,
+		)
+	}
+
+	// The new epoch to be proven in the Bitcoin difficulty chain.
+	newEpoch := currentEpoch + 1
+
+	// Height of the first block of the new epoch.
+	newEpochHeight := newEpoch * bitcoinEpochLength
+
+	// The range of block headers to be pull from the Bitcoin chain should
+	// start `proofLength` blocks before the first block of the new epoch
+	// and end `proofLength`-1 after it.
+	// For example, if the new epoch begins at block 522144 and `proofLength`
+	// is 3, then the range should be [522141, 522146]:
+	// 522141 <- old difficulty target
+	// 522142 <- old difficulty target
+	// 522143 <- old difficulty target
+	// << difficulty retarget >>
+	// 522144 <- new difficulty target (first block of the new epoch)
+	// 522145 <- new difficulty target
+	// 522146 <- new difficulty target
+	firstBlockHeaderHeight := uint(newEpochHeight - proofLength)
+	lastBlockHeaderHeight := uint(newEpochHeight + proofLength - 1)
+
+	// The required range of block headers can be pulled from the Bitcoin
+	// blockchain only if the blockchain height is equal to or greater than
+	// the end of the range.
+	if currentBlockNumber >= lastBlockHeaderHeight {
+		headers, err := bdm.getBlockHeaders(
+			firstBlockHeaderHeight,
+			lastBlockHeaderHeight,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to get block headers from Bitcoin chain [%v]",
+				err,
+			)
+		}
+
+		if err := bdm.chain.Retarget(headers); err != nil {
+			return fmt.Errorf(
+				"failed to submit block headers from range [%v, %v] to "+
+					"the Bitcoin difficulty chain [%v]",
+				firstBlockHeaderHeight,
+				lastBlockHeaderHeight,
+				err,
+			)
+		}
+
+		logger.Infof(
+			"Successfully submitted block headers [%v:%v] to the Bitcoin "+
+				"difficulty chain. The current proven epoch is %v.",
```

Comment:
To be consistent with the codebase can we update all the logged messages to start lowercase?
```suggestion
			"successfully submitted block headers [%d:%d] to the Bitcoin "+
				"difficulty chain; the current proven epoch is %d",
```

@pdyraga do we have any guidelines for log and error messages written down?